tools/offline_log_viewer/kafka.py

```diff
@@ -32,18 +42,98 @@ def decode(self):
+                        "<q", record.value[23 + size:31 + size])[0]
+                    record_dict["value"]["status"] = TxStatus(
+                        struct.unpack("<i", record.value[31 + size:35 +
+                                                         size])[0]).name
+                    record_dict["value"]["timeout_ms"] = struct.unpack(
+                        "<q", record.value[35 + size:43 + size])[0]
+                    record_dict["value"]["last_update_ts"] = struct.unpack(
+                        "<q", record.value[43 + size:51 + size])[0]
+                    record_dict["value"]["partitions"] = []
+                    p_size = struct.unpack("<i", record.value[51 + size:55 +
+                                                              size])[0]
+                    o = 55 + size
+                    for i in range(p_size):
+                        size = struct.unpack("<i", record.value[o:o + 4])[0]
+                        o += 4
+                        partition = {
+                            "ntp": {
+                                "ns": record.value[o:o + size].decode("UTF-8")
+                            }
+                        }
+                        o += size
+                        size = struct.unpack("<i", record.value[o:o + 4])[0]
+                        o += 4
+                        partition["ntp"]["tp"] = {}
+                        partition["ntp"]["tp"]["topic"] = record.value[
+                            o:o + size].decode("UTF-8")
+                        o += size
+                        partition["ntp"]["tp"]["partition"] = struct.unpack(
+                            "<i", record.value[o:o + 4])[0]
+                        o += 4
+                        partition["etag"] = struct.unpack(
+                            "<q", record.value[o:o + 8])[0]
+                        o += 8
+                        record_dict["value"]["partitions"].append(partition)
+                    g_size = struct.unpack("<i", record.value[o:o + 4])[0]
+                    o += 4
+                    record_dict["value"]["groups"] = []
+                    for i in range(g_size):
+                        group = {}
+                        size = struct.unpack("<i", record.value[o:o + 4])[0]
+                        o += 4
+                        group["id"] = record.value[o:o + size].decode("UTF-8")
+                        o += size
+                        group["etag"] = struct.unpack("<q",
+                                                      record.value[o:o + 8])[0]
+                        o += 8
+                        record_dict["value"]["groups"].append(group)
+                elif header["type_name"] == "raft_configuration":
+                    pass
+                elif header["type_name"] == "checkpoint":
+                    pass
+                else:
+                    raise Exception(header["type_name"])
+                yield record_dict
```

Comment:
why not keep this low level record iterator as is, and let the caller decode based on type?

Reply:
I've refactored it similar to `consumer_groups`